### PR TITLE
ENH: Allow refitting HPI and dev_head_t

### PR DESCRIPTION
**Pull request by larsoner**
Sometimes the Neuromag system screws up the initial HPI fitting, which then messes up `dev_head_t` and also any subsequent continuous head position estimation. This PR will fix that. Lots of TODOs in the code but I wanted to get it up to have a place to track progress.

- [x] Get it working with problematic dataset
- [x] Add test to show necessity of `if proj_op is None` addition
- [x] Show that it's essentially a no-op on correct data
- [x] Add `whats_new` entry
- [x] Add elements from problematic dataset to tests
- [x] Show it fails without refitting
- [x] Show it works with refitting
- [x] Ensure all TODOs inline are addressed

### CODE REVIEW

**Comment by larsoner** [comment]
@drammock I think this one is good to go!

**Comment by larsoner** [comment]
I don't see this one as a blocker for 1.11 -- can be an early 1.12 so bumping milestone

File: mne/utils/_testing.py [hunk 1]
```diff
@@ -357,3 +357,27 @@ def _get_suptitle(fig):
...
+    assert isinstance(desired, np.ndarray)
+    assert actual.shape == (4, 4)
+    assert desired.shape == (4, 4)
+    angle, dist = angle_distance_between_rigid(
+        actual, desired, angle_units="deg", distance_units="m"
+    )
+    assert dist <= dist_tol, (
+        f"{1000 * dist:0.3f} > {1000 * dist_tol:0.3f} mm translation"
+    )
+    assert angle <= angle_tol, f"{angle:0.3f} > {angle_tol:0.3f}° rotation"
```
Comment (drammock) [inline mne/utils/_testing.py:419]: might be good for more of these `asserts` to have explanatory text for failures

File: mne/chpi.py [hunk 2]
```diff
@@ -1615,3 +1625,310 @@ def get_active_chpi(raw, *, on_missing="raise", verbose=None):
...
+    gof_limit : float
+        The goodness-of-fit limit to use when choosing which coils to use for refitting.
+    dist_limit : float
+        The distance limit (in meters) to use when choosing which coils to use for
+        refitting.
+    use : int | None
+        The maximum number of coils to use when testing different coil orderings.
+        The default for ``hpifit`` in MEGIN software is 3. Default (None) means to
+        use all coils above ``gof_limit``. Can also be an ndarray of coil indices
+        (0-indexed!) to use, e.g., ``[2, 3, 5]``.
```
Comment (drammock) [inline mne/chpi.py]:
```suggestion
        (0-indexed!) to use, e.g., ``[1, 2, 4]``.
```
if it's 0-indexed, including `5` implies there were 6 coils to begin with. I thought most systems only used 5 coils?

File: mne/chpi.py [hunk 3]
```diff
@@ -1615,3 +1625,310 @@ def get_active_chpi(raw, *, on_missing="raise", verbose=None):
...
+    linearity_limit : float
+        The RMS limit (in meters) to use when checking for colinearity of coils.
```
Comment (drammock) [inline mne/chpi.py]: I think this should say "colinearity of coil locations". On first read, I was assuming this was about coil data (in spite of the "in meters" note earlier in the sentence).  Also maybe the param name should be `colinearity_limit`?

File: mne/chpi.py [hunk 4]
```diff
@@ -1615,3 +1625,310 @@ def get_active_chpi(raw, *, on_missing="raise", verbose=None):
...
+
+    1. Optionally recompute HPI amplitudes (sinusoidal fit for each channel) using
+       :func:`~mne.chpi.compute_chpi_amplitudes`.
+    2. Optionally use HPI amplitudes to fit HPI coil locations using
+       :func:`~mne.chpi.compute_chpi_locs`.
+    3. Optionally determine coil digitization order by testing all permutations
+       for the best goodness of fit between digitized coil locations and
+       (rigid-transformed) fitted coil locations.
+    4. Subselect coils to use for fitting ``dev_head_t`` based on ``gof_limit``,
+       ``dist_limit``, and ``max_use``.
```
Comment (drammock) [inline mne/chpi.py]:
```suggestion
       ``dist_limit``, and ``use``.
```
or, change param name to `max_use`, that might be better actually?

File: mne/chpi.py [hunk 5]
```diff
@@ -1615,3 +1625,310 @@ def get_active_chpi(raw, *, on_missing="raise", verbose=None):
...
+    if amplitudes and not locs:
+        raise ValueError(
+            "If amplitudes is True, locs must also be True (otherwise "
+            "recomputing amplitudes has no effect)"
+        )
+    if locs and not order:
+        raise ValueError(
+            "If locs is True, order must also be True (otherwise "
+            "recomputing locations has no effect)"
+        )
```
Comment (drammock) [inline mne/chpi.py:1758]: I'm +0.5 for adding this info to the docstring as well, just to save diligent readers one cycle of run-error-debug-rerun

File: mne/chpi.py [hunk 6]
```diff
@@ -1615,3 +1625,310 @@ def get_active_chpi(raw, *, on_missing="raise", verbose=None):
...
+    results["used"] = used + 1  # make 1-indexed (can be different length from previous)
+    results["order"][:] = fit_order + 1  # make 1-indexed
+    results["goodness"][:] = hpi_gofs
+    results["good_limit"] = gof_limit
+    results["moments"][:] = fit_locs["moments"][0]
+    results["dist_limit"] = dist_limit
+    del fit_locs
+    for ci, loc in enumerate(hpi_dev):
+        results["dig_points"][ci]["r"][:] = loc.astype(float)
+    # result["moments"] = ndarray, shape (n_coils, 3)
```
Comment (drammock) [inline mne/chpi.py]: cruft?

**Review by drammock — APPROVED** [review]
just a few comments/suggestions/questions.  Didn't look too closely at the tests.

File: mne/utils/_testing.py (hunk 1 above)
Comment (larsoner) [reply mne/utils/_testing.py:419]: I thought the `pytest` default `assert` output should ideally be good enough. But it turned out:
```
assert actual.shape == (4, 4)
```
Didn't say anything. BUT! I discovered:

https://docs.pytest.org/en/stable/how-to/assert.html#assertion-introspection-details

can make it say something like this (if I change it to `(3, 4)` just to get it to fail, which seems clear enough to me given the traceback
```
mne/tests/test_chpi.py:976: in test_refit_hpi_locs_problematic
    assert_trans_allclose(
E   assert (4, 4) == (3, 4)
E     
E     At index 0 diff: 4 != 3
E     Use -v to get more diff
```
so knowing that it overwrites some of it, I can at least add a little less text. Will push a commit shortly

File: mne/chpi.py (hunk 4 above)
Comment (larsoner) [reply mne/chpi.py]: Well it can be a list of coils to use, too, which is why I changed it from `max_use` (when initially I only allowed setting it as a max count) to `use` (when I also allowed it to be a list of int)

**Comment by larsoner** [comment]
Thanks @drammock I think I addressed the comments! I'll mark for merge-when-green to sneak it in for 1.11. I think it should be good enough for people to try
